pallets/token-mux/src/lib.rs

```diff
@@ -0,0 +1,328 @@
+// Copyright 2023 Centrifuge Foundation (centrifuge.io).
+//
+// This file is part of the Centrifuge chain project.
+// Centrifuge is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version (see http://www.gnu.org/licenses).
+// Centrifuge is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+// TODO(william): Add pallet description
+
+#[cfg(test)]
+mod mock;
+
+#[cfg(test)]
+mod tests;
+
+#[cfg(feature = "runtime-benchmarks")]
+mod benchmarking;
+
+pub mod weights;
+
+pub use cfg_traits::TokenSwaps;
+pub use pallet::*;
+pub use weights::WeightInfo;
+
+#[frame_support::pallet]
+pub mod pallet {
+
+	use cfg_traits::{OrderDetails, OrderRatio, TokenSwaps};
+	use cfg_types::{
+		orders::MuxSwap,
+		tokens::{CustomMetadata, LocalAssetId},
+	};
+	use frame_support::{
+		pallet_prelude::{DispatchResult, *},
+		traits::{
+			fungibles,
+			fungibles::Mutate,
+			tokens::{Fortitude, Precision, Preservation},
+		},
+		PalletId,
+	};
+	use frame_system::pallet_prelude::{OriginFor, *};
+	use orml_traits::asset_registry::{self, Inspect as _};
+	use sp_arithmetic::FixedPointOperand;
+	use sp_runtime::traits::{AccountIdConversion, EnsureFixedPointNumber, One};
+
+	use super::*;
+
+	type AccountIdFor<T> = <T as frame_system::Config>::AccountId;
+	pub type BalanceFor<T> =
+		<<T as Config>::Tokens as fungibles::Inspect<AccountIdFor<T>>>::Balance;
+	pub type CurrencyFor<T> =
+		<<T as Config>::Tokens as fungibles::Inspect<AccountIdFor<T>>>::AssetId;
+	pub type OrderIdFor<T> = <<T as Config>::Swaps as TokenSwaps<AccountIdFor<T>>>::OrderId;
+	pub type RatioFor<T> = <<T as Config>::Swaps as TokenSwaps<AccountIdFor<T>>>::Ratio;
+
+	const STORAGE_VERSION: StorageVersion = StorageVersion::new(0);
+
+	#[pallet::pallet]
+	#[pallet::storage_version(STORAGE_VERSION)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config
+	where
+		CurrencyFor<Self>: From<LocalAssetId> + TryInto<LocalAssetId>,
```

Comment:
Currently, all related with currencies are a bit mixed in the codebase, but theoretically, `LocalAssetId` is a runtime type. If we're abstracting the `CurrencyId,` we should also abstract `LocalAssetId`. Maybe makes sense to have this type under another associated type in `Config`:

```rust
trait Config {
    type LocalAssetId;
    type CurrencyId: From<Self::LocalAssetId> + TryInto<Self::LocalAssetId>
}
```

Reply:
Since the `cfg_types::tokens::LocalAssetId` is baked into the `CustomMetadata`, we still need to restrict the associated `LocalAssetId` Config type to impl `From<cfg_types::tokens::LocalAssetId>`. Not sure if this makes sense to be refactored then.

Reply:
Mmm... I see 🤔. Our `CustomMetadata` forces us to bring a lot of types from the runtime into testing. Maybe we should abstract the `CustomeMetadata` somehow. But sure, it's not work for this PR.

Then, as you said, it does not make sense to refactor it. Use `LocalAssetId` directly.